### ISSUE TITLE
UX for edit modes, POIs, and routes

### ISSUE THREAD

**Issue by zstadler** [IsraelHikingMap/Site#537]
This is a summary of the expected user experience in the different edit modes with respect to private routes and POIs, routes and POI from the public layers, and search results.

#### Edit modes
<image width="32px" src="https://user-images.githubusercontent.com/1304610/32265198-b852c8c6-beea-11e7-8888-ae39a4f2014a.png"> POI modify mode, previously referred to as "Read-only mode"    
<image width="32px" src="https://user-images.githubusercontent.com/1304610/32265050-0c29b9a6-beea-11e7-9509-eebbaa3bf1de.png"> POI create mode    
<image width="32px" src="https://user-images.githubusercontent.com/1304610/32264875-64f01054-bee9-11e7-8fff-3f8aac16ec46.png"> Route edit mode    

#### Cursor attachment
Depending on the edit mode, the cursor may attach to certain positions on the map where a significant entity exists. When the cursor is attached to an entity, the UX will relate to it. Priority of different entities to cursor attachment allows useful user experience when multiple entities share the same location, such as a public POI added to a private route layer or a route going through a POI on an OSM road.

#### Private route layers
<image width="200px" src="https://user-images.githubusercontent.com/1304610/32267562-5502d734-bef4-11e7-80ba-8f62ca37d9a4.png">

A layer may include up-to one private track and any number of private POIs. 

* If a layer is active, edit operations can be done on its track and POIs
* If a layer is inactive, its track and POIs can be seen on the map, but not edited
* if a layer is hidden, its track and POIs are not seen on the map

#### Public POIs, routes, and the last search result
<image width="200px" src="https://user-images.githubusercontent.com/1304610/32268170-93d1c824-bef6-11e7-8b16-6e61193a528e.png">
When visible, show icons for public POIs and routes. At this time, the user interface allows a logged-in user to vote on all public POIs and routes. In addition there are some basic edit capabilities for OSM POIs and routes:

* OSM public POIs - create, add pictures, modify title, description, and location. At this time all public POIs are provided by OSM.
* OSM public routes - Edit the description and add pictures.

When a search result is selected, the adds an icon to the map. This last search result is considered a public POI or route as appropriate.

### User experience of mouse operations
Edit mode can only be changed by clicking its the edit buttons. There is no explicit change of edit mode.

#### [POI modify mode](edit-modes)
* Cursor attachment is done in this priority order:
  * POIs of the [active layer](private-route-layers)
  * Public POIs 
  * Icons of public routes
* A click on a private POI edits the POI
* Click-and-drag pans the map
* A click on a public POIs or routes shows their information dialog

#### [POI create mode](edit-modes)
* Cursor attachment is done in this priority order:
  * POIs of the [active layer](private-route-layers)
  * Public POIs 
  * Icons of public routes
* A click on a private POI enabled editing of the POI
* A click on a public POIs or routes shows their information dialog.
* Click-and-drag on a private POI moves it
* Click-and-drag elsewhere pans the map

#### [Route edit mode ](edit-modes)
* If technically possible, cursor attachment is done in this priority order:
  * POIs of the [active layer](private-route-layers)
  * Public POIs
  * Route points
  * Segments of the route
  * Roads locations.
* Clicks
  * Click location is determined by the cursor attachment, if any
  * The first click creates the route's start point.
  * A click on an existing route point opens the pop-up to allow removal
  * Any other click extends the route by adding a point to the end of route, which is not a POI.
* Click-and-drag
  * Click-and-drag on an existing route point moves it
  * Click-and-drag on an a location along the route creates a new point and moves it
  * Click-and-drag elsewhere pans the map

**Comment by HarelM** [comment]
I wanted to start implementing this and looked at the code.
Also I found this issue to be related: #183
My main concern here is that we add complexity instead of ease-of-use:
Currently a user in edit mode can edit every entity in the route (both POI and the route itself).
This feature will introduce a need for a user to switch between POI edit and route edit and I'm not sure this is a positive change.

**Comment by zstadler** [comment]
One of the primary triggers for this definition change was the following use case:
- Add several POIs to a route layer, either manually or using the POI search.
- Create a route that will go through these POIs in some order.

Without a clear separation into "POI edit" and "Route edit" modes, this scenario will be difficult to support.

**Comment by zstadler** [comment]
@HarelM We agreed about the need to separate the "POI edit" and "Route edit" modes.

**Comment by HarelM** [comment]
@zstadler There's an initial version [here](https://israelhiking.osm.org.il/beta/) with some of the changes.
I still think there's a good reason to keep read-only state especially if we're going to implement #590 - meaning that a private POI will have a label (tooltip in leaflet terms) and when clicking on that POI a popup will be opened and the user can see the image and the text, I thought about also adding an edit button which will change the state to POI add/edit mode (kind of a shortcut).

In general, there are two aspects I can play with:
1. Snapping - to which entity to snap to, I've implemented different logic for POI and route editing - this can be seen by the hover marker - it is only relevant for editing modes when adding a point.
2. ZIndexOffset - I would like to avoid touching this as much as possible but I have set it to 1000 for all route points when editing a route, which means that the route markers are always the first to get a click, which I think makes sense when editing a route.
Bottom line: The behavior can be defined to each state separately which allows nice flexibility. 

Let me know if you encounter any issues.

**Comment by zstadler** [comment]
@HarelM 

Thanks for the beta version. I'm trying it.

The terminology presented above has 3 edit modes, namely:

1. POI modify mode, previously referred to as "Read-only mode"
2. POI create mode
3. Route edit mode 

There is no request to remove the "Read-only mode". The expected UX in this mode was described, including the previous ability to modify an existing private POI. Accordingly, the name of the mode was updated. I therefore do not understand the paragraph that starts with:

> I still think there's a good reason to keep read-only state

Are you suggesting (a) to add a 4th edit mode, (b) to have a different name for the "POI modify mode", or (c) to define a different UX in specific aspects of the "POI modify mode"?

**Comment by HarelM** [comment]
I propose (c) change the UX such that no changes can be done in POIs and routes if none of the edit buttons is active. This will enable showing detailed POI information when a POI is clicked rather than directly entering edit mode for the POI
As a result, the name of the mode should be "Read Only mode", as it was previously referred to.

**Comment by HarelM** [comment]
I have published this code, I'm keeping this issue open until we decide this is the required behavior. Once we do (or no objection is raised) I'll merge the changes to the master branch and close this issue.

**Comment by HarelM** [comment]
Closing this as main development was done. please open separate issues if stuff are not behaving as expected.
